Winton.Extensions.Threading.Actor/Internal/ActorTask.cs

```diff
@@ -1,24 +1,65 @@
-using System;
+﻿using System;
+using System.Diagnostics;
+using System.Threading;
 using System.Threading.Tasks;
 
 namespace Winton.Extensions.Threading.Actor.Internal
 {
-    internal static class ActorTask
+    internal sealed class ActorTask
     {
-        public static bool IsActorTask(this Task self)
+        [ThreadStatic]
+        private static CancellationTokenSource _currentCanceller;
+
+        [ThreadStatic]
+        private static ActorTaskTraits _currentActorTaskTraits;
+
+        private CancellationTokenSource _canceller;
+
+        public ActorTask(Task task)
+        {
+            Debug.Assert(_currentCanceller != null);
+
+            Task = task;
+            Traits = _currentActorTaskTraits;
+            _canceller = _currentCanceller;
+
+            _currentCanceller = null;
+            _currentActorTaskTraits = ActorTaskTraits.None;
```

Comment:
The values are only intended for the construction of a single `ActorTask` so I reset them after each construction. `ActorTaskFactory` sets these then queues a `Task` (via `Task.Factory.StartNew`) on the `ActorTaskScheduler` which will result in the synchronous call to `QueueTask` that constructs `ActorTask`.

It's a way of injecting data the scheduler needs. Previously I used `Task.AsyncState` for this but this is more performant for various reasons.

Reply:
From my naive understanding, would it not be possible to pass those values to the `ActorTask` constructor from the `ActorTaskFactory`? That would avoid the shared construction state across different instances.

Reply:
No. To do that I would need `ActorTask` to derive from `Task` so that in `ActorTaskFactory` I could do:

```
var actorTask = new ActorTask(...);
actorTask.Start(_scheduler);
```

rather than calling `Task.Factory.StartNew`. However, directly instantiating a `Task` then starting it rather than calling `Task.Run` or `Task.Factory.StartNew` is a bad idea (there's a [Stephen Cleary blog](http://blog.stephencleary.com/2014/05/a-tour-of-task-part-1-constructors.html) about it). You can get race conditions with the passed-in cancellation token being cancelled before `Start` is called on the `Task`, so that `Start` will throw an `InvalidOperationException`. I actually went this route initially and hit this race in tests. You can try to come up with a mechanism to get round it but it's nasty and not fool-proof. This injection via thread statics is the simplest alternative given I therefore need to use a vanilla `Task` outside the scheduler.